### PR TITLE
fix: setupFilesAfterEnv property cannot be set (#346)

### PR DESCRIPTION
**Pull request by michaeljota**
This will allow to declare properties that exist in the default configuration as array, and are not mean to be override. This is done by declaring a whitelist of properties that should concat their values. By using a helper function with `mergeWith` we can concat only those properties.

### CODE REVIEW

**Comment by meltedspark** [comment]
Looks good. Why do you think it makes sense to concat AST transformers as well? What if someone actually wants to override it?

**Comment by michaeljota** [comment]
I think is better to prevent an undesired behavior. I understand that there may be a use case where you would actually like to replace this entirely, but as I see it, most of the time if you would like to add something on top of the default value, just as setupFilesAfterEnv.

File: packages/jest/src/jest-configuration-builder.ts [hunk 1]
```diff
@@ -1,8 +1,38 @@
 import { Path, resolve } from "@angular-devkit/core";
-import { merge } from 'lodash';
+import { mergeWith } from 'lodash';
 import { CustomConfigResolver } from "./custom-config.resolver";
 import { DefaultConfigResolver } from "./default-config.resolver";
 
+/**
+ * A whitelist of property names that are mean to be concat.
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: *meant

File: packages/jest/src/jest-configuration-builder.ts [hunk 2]
```diff
@@ -1,8 +1,38 @@
...
+  // From ts-jest config
+  'astTransformers'
+];
+
+/**
+ * This function checks witch properties should be concat. Early return will
+ * merge the data as lodash#merge would do it. 
+ */
+function concatArrayProperties(objValue: unknown, srcValue: unknown, property: string) {
+  if (!ARRAY_PROPERTIES_TO_CONCAT.includes(property)) {
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: Can be united

File: packages/jest/src/jest-configuration-builder.ts [hunk 3]
```diff
@@ -1,8 +1,38 @@
...
+    return;
+  }
+
+  if (!Array.isArray(objValue)) {
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: Can be united with the previous if statement

File: packages/jest/src/jest-configuration-builder.ts [hunk 4]
```diff
@@ -1,8 +1,38 @@
...
+    return;
+  }
+  
+  if (!Array.isArray(srcValue)) {
+    // This will make Jest to throw, as it is expecting an array, instead of silently fail.
+    return srcValue;
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: I'd throw an explicit error here: "Tried to concat non-array property". Moreover the behavior should be the same for object value as well. If I tried to merge array value into non-array value while I listed it in the ARRAY properties I probably mistaken.

File: packages/jest/src/jest-configuration-builder.ts [hunk 5]
```diff
@@ -24,12 +54,7 @@ export class JestConfigurationBuilder {
   buildConfiguration(projectRoot: Path, workspaceRoot: Path, configPath: string = 'jest.config.js'): any {
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: No need in default argument anymore, since you're using buildConfiguration function

File: packages/jest/src/jest-configuration-builder.ts (hunk 3 above)
Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: Sure. I miss that.

File: packages/jest/src/jest-configuration-builder.ts (hunk 4 above)
Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: But this would have an inconsistent behavior for the properties we don't concat, and replace instead. If we leave Jest to throw, it would throw the same error, regardless is a property we need to concat or not.

Comment (meltedspark) [reply packages/jest/src/jest-configuration-builder.ts]: We don't want the error to be thrown if it's not a property we want to concat. We do want to throw an error if it is a property we want to concat but not an array. So the inconsistency is fine, what am I missing?

Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: Nothing. Just that I though it would be better to let Jest to throw, as it is currently happening. But I'll update it.

File: packages/jest/src/jest-configuration-builder.ts (hunk 1 above)
Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: I update this as well. Thanks.

**Comment by michaeljota** [comment]
A though about this. What if the user configure some of this on the angular.json and some of this on the jest.config? .-.

**Comment by meltedspark** [comment]
It is not necessarily related to this pull request, but if you look [here](https://github.com/meltedspark/angular-builders/blob/master/packages/jest/src/index.ts) you'll see that the configuration is provided to Jest as one of the CLI arguments. So if an argument that already exists in configuration is provided as a separate CLI option the behavior would be the same as default Jest behavior. Need to verify that but I think it would override the config param.

File: packages/jest/src/jest-configuration-builder.ts [hunk 6]
```diff
@@ -18,17 +18,12 @@ const ARRAY_PROPERTIES_TO_CONCAT = [
  * merge the data as lodash#merge would do it. 
  */
 function concatArrayProperties(objValue: unknown, srcValue: unknown, property: string) {
-  if (!ARRAY_PROPERTIES_TO_CONCAT.includes(property)) {
+  if (!ARRAY_PROPERTIES_TO_CONCAT.includes(property) || !Array.isArray(objValue)) {
```
Comment (meltedspark) [inline packages/jest/src/jest-configuration-builder.ts]: I think of `objValue` is not an array it should throw as well, no? It still means that we're trying to concat a property which is not an array, meaning that we've put it in the list by mistake.

Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: I think you are right. But that should be catch with unit testing. So, I will remove that from the code.

Comment (michaeljota) [reply packages/jest/src/jest-configuration-builder.ts]: Updated.

**Comment by michaeljota** [comment]
Sorry for the delay. I updated the PR to address the latest comment.

**Comment by meltedspark** [comment]
@michaeljota It is still inconsistent. It's one of the two:
1. We say that we are sure that the properties we put in `ARRAY_PROPERTIES_TO_CONCAT` are in fact array properties (and we cover it with UT as you said). In this case there is no need to validate either source value or object value for being an array, because, well, we know that if it's in `ARRAY_PROPERTIES_TO_CONCAT` it is an array.
2. We are not sure about this. In this case there are two possible mistakes: 
    - We're trying to contact array to non-array property
    - We're trying to contact object to an array property
   Both mistakes are possible so we should validate both objValue and srcValue for being an array and throw an error if they are not.

In conclusion: we should either validate both values or none. Validating one of them gives us inconsistent behavior.

**Comment by michaeljota** [comment]
@meltedspark We are validating what's coming from the user input, as we would be validating our own input twice, and this seems like a mistake.

**Comment by michaeljota** [comment]
If you don't want to verify input from the user, then I'll updated. This is actually not necessary, but only made as the user should use the same input schema as Jest is intended.

**Comment by michaeljota** [comment]
@meltedspark with the latest changes that you did, is this ready to merge? Thank you.

**Comment by meltedspark** [comment]
Sure, just wanted to make sure it's fine by you. It is your PR after all.

**Comment by michaeljota** [comment]
Oh! Thanks. Yes. I ask you to do something like what you did.

**Comment by just-jeb** [comment]
@all-contributors please add @michaeljota for code
